src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestLocalCallables1.ll

```diff
@@ -28,84 +30,77 @@ exit__1:                                          ; preds = %header__1
-  %17 = call %Tuple* @__quantum__rt__tuple_create(i64 ptrtoint ({ %String* }* getelementptr ({ %String* }, { %String* }* null, i32 1) to i64))
-  %18 = bitcast %Tuple* %17 to { %String* }*
-  %19 = getelementptr inbounds { %String* }, { %String* }* %18, i32 0, i32 0
-  store %String* %16, %String** %19, align 8
-  %20 = call %Tuple* @__quantum__rt__tuple_create(i64 ptrtoint ({ %String*, { i64, double }* }* getelementptr ({ %String*, { i64, double }* }, { %String*, { i64, double }* }* null, i32 1) to i64))
-  call void @__quantum__rt__callable_invoke(%Callable* %fct, %Tuple* %17, %Tuple* %20)
-  %21 = bitcast %Tuple* %20 to { %String*, { i64, double }* }*
-  %22 = getelementptr inbounds { %String*, { i64, double }* }, { %String*, { i64, double }* }* %21, i32 0, i32 0
-  %str = load %String*, %String** %22, align 8
-  %23 = getelementptr inbounds { %String*, { i64, double }* }, { %String*, { i64, double }* }* %21, i32 0, i32 1
-  %24 = load { i64, double }*, { i64, double }** %23, align 8
-  %25 = getelementptr inbounds { i64, double }, { i64, double }* %24, i32 0, i32 1
-  %val = load double, double* %25, align 8
+  %11 = call %String* @__quantum__rt__string_create(i8* getelementptr inbounds ([1 x i8], [1 x i8]* @0, i32 0, i32 0))
+  %12 = call { %String*, { i64, double }* }* @Microsoft__Quantum__Testing__QIR__ReturnTuple__body(%String* %11)
+  %13 = getelementptr inbounds { %String*, { i64, double }* }, { %String*, { i64, double }* }* %12, i32 0, i32 0
+  %str = load %String*, %String** %13, align 8
+  %14 = getelementptr inbounds { %String*, { i64, double }* }, { %String*, { i64, double }* }* %12, i32 0, i32 1
+  %15 = load { i64, double }*, { i64, double }** %14, align 8
+  %16 = getelementptr inbounds { i64, double }, { i64, double }* %15, i32 0, i32 1
+  %val = load double, double* %16, align 8
   call void @__quantum__rt__string_update_reference_count(%String* %str, i32 1)
-  %26 = call %Tuple* @__quantum__rt__tuple_create(i64 ptrtoint ({ %String*, double }* getelementptr ({ %String*, double }, { %String*, double }* null, i32 1) to i64))
-  %27 = bitcast %Tuple* %26 to { %String*, double }*
-  %28 = getelementptr inbounds { %String*, double }, { %String*, double }* %27, i32 0, i32 0
-  %29 = getelementptr inbounds { %String*, double }, { %String*, double }* %27, i32 0, i32 1
-  store %String* %str, %String** %28, align 8
-  store double %val, double* %29, align 8
+  %17 = call %Tuple* @__quantum__rt__tuple_create(i64 ptrtoint ({ %String*, double }* getelementptr ({ %String*, double }, { %String*, double }* null, i32 1) to i64))
+  %18 = bitcast %Tuple* %17 to { %String*, double }*
+  %19 = getelementptr inbounds { %String*, double }, { %String*, double }* %18, i32 0, i32 0
+  %20 = getelementptr inbounds { %String*, double }, { %String*, double }* %18, i32 0, i32 1
+  store %String* %str, %String** %19, align 8
+  store double %val, double* %20, align 8
+  call void @__quantum__rt__callable_make_adjoint(%Callable* %9)
```

Comment:
It is unfortunate the callable management code is still generated even though the callables are now never invoked. Hopefully we can find a future optimization that recoginizes the callables are never really used and skip outputting any of this code, because LLVM doesn't know these functions are side effect free and therefore won't optimize them away (from the LLVM pass perspective, the callabes are used because they are manipulated by the ref counting and function runtime functions).

Reply:
I think there are a couple of LLVM attributes we should look into that may make it clear that these functions are side effect free. That said, the caching is already avoiding to apply functors when possible - just in some cases it is not clear that it is possible. The functors are applied lazily and/but accessing the .Value property of the CallableValue triggers their application (because I have to assume that I need to apply them at this point).